File: Neetcode/ArrsHashes/arrprodwoself.py
# https://neetcode.io/problems/products-of-array-discluding-self
# https://leetcode.com/problems/product-of-array-except-self/
from types import List


class Solution(object):
    def productExceptSelf2(self, nums: List[int]) -> List[int]:
        zeros = -1
        prod = 1
        for i, n in enumerate(nums):
            if n == 0:
                if zeros != -1:
                    return [0] * len(nums)
                else:
                    zeros = i
            else:
                prod *= n
        if zeros == -1:
            return [prod // n for n in nums]
        else:
            return [0] * zeros + [prod] + [0] * (len(nums) - zeros - 1)

    def productExceptSelf(self, nums):
        prod = 1
        prod2 = 1
        zeros = 0

        for num in nums:
            prod *= num
            if num:
                prod2 *= num
            else:
                zeros += 1

        for i in range(len(nums)):
            if nums[i]:
                nums[i] = prod / nums[i]
            elif zeros == 1:
                nums[i] = prod2
            else:
                nums[i] = 0
        return nums

File: Neetcode/ArrsHashes/arrprodwoself.cpp
// https://leetcode.com/problems/product-of-array-except-self/

#include <vector>
using namespace std;

class Solution {
 public:
  vector<int> productExceptSelf(vector<int>& nums) {
    int lprod = 1;
    int rprod = 1;
    int n = nums.size();

    vector<int> ret(n, 1);
    for (int i = 0; i < n; i++) {
      ret[i] *= lprod;
      ret[n - i - 1] *= rprod;
      lprod *= nums[i];
      rprod *= nums[n - i - 1];
    }
    return ret;
  }
};

class Solution {
 public:
  vector<int> productExceptSelf(vector<int>& nums) {
    int nprod = 1;
    int zero = -1;
    int n = nums.size();
    for (int i = 0; i < n; i++) {
      if (nums[i] == 0) {
        if (zero != -1) {
          zero = -2;
        } else {
          zero = i;
        }
      } else {
        nprod *= nums[i];
      }
    }

    vector<int> ret(nums.size(), 0);
    if (zero == -2) {
      return ret;
    }
    if (zero == -1) {
      for (int i = 0; i < n; i++) {
        ret[i] = nprod / nums[i];
      }
      return ret;
    }
    ret[zero] = nprod;
    return ret;
  }
};